File: island_sim/src/simulator/phases/trade.py
"""Trade phase: port trade and tech diffusion."""

from __future__ import annotations

import numpy as np

from ...types import WorldState
from ..params import SimParams
from ..world import distance


def run_trade(state: WorldState, params: SimParams, rng: np.random.Generator) -> None:
    """Execute the trade phase between ports."""
    # Find all alive ports
    ports = [s for s in state.settlements if s.alive and s.has_port]

    # Track which pairs already traded this turn
    traded: set[tuple[int, int]] = set()

    for idx_a in range(len(ports)):
        for idx_b in range(idx_a + 1, len(ports)):
            port_a = ports[idx_a]
            port_b = ports[idx_b]

            dist = distance(port_a.x, port_a.y, port_b.x, port_b.y)
            if dist > params.trade_range:
                continue

            # Skip if factions are at war (raided each other this turn)
            if port_a.owner_id != port_b.owner_id:
                war_key = (min(port_a.owner_id, port_b.owner_id),
                           max(port_a.owner_id, port_b.owner_id))
                if war_key in state.war_pairs:
                    continue

            pair_key = (
                min((port_a.x, port_a.y), (port_b.x, port_b.y)),
                max((port_a.x, port_a.y), (port_b.x, port_b.y)),
            )
            if pair_key in traded:
                continue
            traded.add(pair_key)

            _execute_trade(port_a, port_b, params)
            _diffuse_tech(port_a, port_b, params)


def _execute_trade(port_a, port_b, params: SimParams) -> None:
    """Exchange food and wealth between two ports."""
    # Both parties benefit from trade
    port_a.food += params.trade_food_gain
    port_b.food += params.trade_food_gain
    port_a.wealth += params.trade_wealth_gain
    port_b.wealth += params.trade_wealth_gain


def _diffuse_tech(port_a, port_b, params: SimParams) -> None:
    """Technology diffuses from higher-tech to lower-tech partner."""
    if port_a.tech_level > port_b.tech_level:
        diff = port_a.tech_level - port_b.tech_level
        port_b.tech_level += diff * params.tech_diffusion_rate
    elif port_b.tech_level > port_a.tech_level:
        diff = port_b.tech_level - port_a.tech_level
        port_a.tech_level += diff * params.tech_diffusion_rate

File: island_sim/rust/src/phases/trade.rs
use crate::params::SimParams;
use crate::types::*;
use crate::world::distance;

pub fn run_trade(state: &mut WorldState, params: &SimParams) {
    // Find all alive port indices
    let port_indices: Vec<usize> = state.settlements.iter()
        .enumerate()
        .filter(|(_, s)| s.alive && s.has_port)
        .map(|(i, _)| i)
        .collect();

    for i in 0..port_indices.len() {
        for j in (i + 1)..port_indices.len() {
            let idx_a = port_indices[i];
            let idx_b = port_indices[j];

            let dist = distance(
                state.settlements[idx_a].x, state.settlements[idx_a].y,
                state.settlements[idx_b].x, state.settlements[idx_b].y,
            );
            if dist > params.trade_range { continue; }

            // Trade
            state.settlements[idx_a].food += params.trade_food_gain;
            state.settlements[idx_b].food += params.trade_food_gain;
            state.settlements[idx_a].wealth += params.trade_wealth_gain;
            state.settlements[idx_b].wealth += params.trade_wealth_gain;

            // Tech diffusion
            let tech_a = state.settlements[idx_a].tech_level;
            let tech_b = state.settlements[idx_b].tech_level;
            if tech_a > tech_b {
                state.settlements[idx_b].tech_level += (tech_a - tech_b) * params.tech_diffusion_rate;
            } else if tech_b > tech_a {
                state.settlements[idx_a].tech_level += (tech_b - tech_a) * params.tech_diffusion_rate;
            }
        }
    }
}
